streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,1123 @@
+
+        streams.setUncaughtExceptionHandler((t, e) -> {
+            if (uncaughtException != null) {
+                e.printStackTrace(System.err);
+                fail("Should only get one uncaught exception from Streams.");
+            }
+            uncaughtException = e;
+        });
+
+        return streams;
+    }
+
+    private void waitForStateTransition(final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed,
+                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> expected)
+        throws Exception {
+
+        try {
+            waitForCondition(
+                () -> observed.equals(expected),
+                MAX_WAIT_TIME_MS,
+                "Client did not startup on time."
+            );
+        } catch (final AssertionError error) {
+            final AssertionError newError = new AssertionError("Client transitions: " + observed);
+            newError.addSuppressed(error);
+            throw newError;
+        }
+    }
+
+    private void waitForStateTransition(final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed1,
+                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> expected1,
+                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed2,
+                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> expected2)
+        throws Exception {
+
+        try {
+            waitForCondition(
+                () -> observed1.equals(expected1) && observed2.equals(expected2),
+                MAX_WAIT_TIME_MS,
+                "Clients did not startup and stabilize on time."
+            );
+        } catch (final AssertionError error) {
+            final AssertionError newError = new AssertionError("Client transitions: " +
+                "\n  client-1 transitions: " + observed1 +
+                "\n  client-2 transitions: " + observed2);
+            newError.addSuppressed(error);
+            throw newError;
+        }
+    }
+
+    private List<KeyValue<Long, Long>> prepareData(final long fromInclusive,
+                                                   final long toExclusive,
+                                                   final Long... keys) {
+        final List<KeyValue<Long, Long>> data = new ArrayList<>();
+
+        for (final Long k : keys) {
+            for (long v = fromInclusive; v < toExclusive; ++v) {
+                data.add(new KeyValue<>(k, v));
+            }
+        }
+
+        return data;
+    }
+
+    private void writeInputData(final List<KeyValue<Long, Long>> records) {
+        final Properties config = TestUtils.producerConfig(
+            CLUSTER.bootstrapServers(),
+            LongSerializer.class,
+            LongSerializer.class
+        );
+        config.setProperty(ProducerConfig.PARTITIONER_CLASS_CONFIG, KeyPartitioner.class.getName());
```

Comment:
With 4 partitions, we need a custom partitioner to make sure we write the 4 different keys into 4 different partitions -- the default partitioner would only write data to 2 partitions (this behavior, ie, empty partitions vs. non-empty partitions, seems to be related to the bug when the test fails -- ensuring that data is written into all partitions avoids the issue).